cabal-install/src/Distribution/Client/Utils.hs

```diff
@@ -488,3 +490,30 @@ safeRead s
 --   This is [Agda.Utils.List.hasElem](https://hackage.haskell.org/package/Agda-2.6.2.2/docs/Agda-Utils-List.html#v:hasElem).
 hasElem :: Ord a => [a] -> a -> Bool
 hasElem xs = (`Set.member` Set.fromList xs)
+
+-- True if x occurs before y
+occursOnlyOrBefore :: (Eq a) => [a] -> a -> a -> Bool
+occursOnlyOrBefore xs x y = case (elemIndex x xs, elemIndex y xs) of
+                       (Just i, Just j) -> i < j
+                       (Just _, _) -> True
+                       _ -> False
+
+giveRTSWarning :: String -> String
+giveRTSWarning "run" = "Your RTS options are applied to cabal, not the "
+               ++ "executable. Use '--' to separate cabal options from your "
+               ++ "executable options. For example, use 'cabal run -- +RTS -N "
+               ++ "to pass the '-N' RTS option to your executable."
+giveRTSWarning "test" = "Some RTS options were found standalone, "
+               ++ "which affect cabal and not the binary. "
+               ++ "Please note that RTS inside proper options suffices "
+               ++ "if your goal is to affect the tested binary. "
+               ++ "For example, use 'cabal test --test-options=\"+RTS -N\"' "
```

Comment:
For silly reasons to do with shells it's in general better to do \"cabal test --test-options=\'+RTS -N'\", because the user can then use double quotes inside the options (and the other way around it doesn't work).